app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+        self.planning_svc = planning_svc
+        self.stopping_conditions = stopping_conditions
+        self.stopping_condition_met = False
+        self.state_machine = ['bayes_state']
+        self.next_bucket = 'bayes_state'
+        self.matrix_past_links = None
+        self.min_link_data = min_link_data
+        self.decision_logging = decision_logging
+        self.min_prob_link_success = (99.0-self.operation.visibility)/(100)
+        self.links_executed = 0
+        if self.decision_logging:
+            self.planning_svc.log.debug("Operation started")   
+
+    async def execute(self):
+        await self.planning_svc.execute_planner(self)
+    
+    async def bayes_state(self):
+        """
+        Update past operational data matrix if needed then run top priority link for each agent until operation concluded
+        """
+        if self.links_executed % 10 == 0:
+            past_operation_data = await self.planning_svc.get_service('data_svc').locate('operations')
+            self.matrix_past_links = await self._build_past_links_matrix(past_operation_data)
+
+        links_to_use = []
+        for agent in self.operation.agents:
+            possible_agent_links = await self.planning_svc.get_links(operation=self.operation, agent=agent)
+            next_link = await self._get_best_link(possible_agent_links)
+            if next_link:
+                links_to_use.append(await self.operation.apply(next_link))
+        if links_to_use:
+            # Each agent will run the next available step.
+            await self.operation.wait_for_links_completion(links_to_use)
+            self.links_executed += 1
+        else:
+            if self.decision_logging:
+                self.planning_svc.log.debug("Operation concluded")
+            self.next_bucket = None
+            self.stopping_condition_met = True
+
+    async def _get_best_link(self, links : List[Link]) -> Union[Link, None]:
+        """
+        Selects the best link from links or None if all links don't match qualifications. 
+
+        :param links: list of links that agent could execute
+        """
+        dict_link_index_to_prob_success = dict()
+        dict_link_index_to_num_observations = dict()
+        links_insufficient_data = []
+
+        for index in range(len(links)):
+            cur_link = links[index]
+            # NOTE: link_query_features allows customizing which of 16 features to query by
+            # current selection of features with broad scope:
+            link_query_features = {
+                "Ability_ID": str(cur_link.ability.ability_id),
+                "Link_Facts": self._get_useful_facts(cur_link),
+                "Executor_Platform": str(cur_link.executor.platform)
+            }
+            prob_link_success, num_link_observations = self._get_link_success_probability(link_query_features)
+            if prob_link_success == None or str(cur_link.ability.ability_id) == "300157e5-f4ad-4569-b533-9d1fa0e74d74":
```

Comment:
This logic may be too specific for just one ability. One other option is that you could have the planner config include a list of ability IDs that should all be handled this way, which would allow for a user to specify which ones to use and would keep this specific check out of the code in case the ability ID of `compress staged directory` ever changes.